File: model/src/main/scala/ru/mipt/acsl/decode/model/domain/impl/registry/RegistryImpl.scala
package ru.mipt.acsl.decode.model.domain.impl.registry

import ru.mipt.acsl.decode.model.domain.impl.naming.{ElementName, Fqn, Namespace}
import ru.mipt.acsl.decode.model.domain.proxy.{DecodeProxyResolver, ExistingElementsProxyResolver, PrimitiveAndGenericTypesProxyResolver}
import ru.mipt.acsl.decode.model.domain.naming.ElementName

import scala.collection.immutable

/**
  * @author Artem Shein
  */
private[decode] class RegistryImpl(val name: ElementName, resolvers: DecodeProxyResolver*) extends Registry {
  if (Fqn.SystemNamespace.size != 1)
    sys.error("not implemented")

  var rootNamespaces: immutable.Seq[Namespace] = immutable.Seq.empty

  val proxyResolvers = resolvers.to[immutable.Seq]

  def this() = this(ElementName.newFromMangledName("GlobalRegistry"),
    new ExistingElementsProxyResolver(),
    new PrimitiveAndGenericTypesProxyResolver())

}

File: model/src/main/java/ru/mipt/acsl/decode/model/registry/RegistryImpl.java
package ru.mipt.acsl.decode.model.registry;

import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Fqn;
import ru.mipt.acsl.decode.model.naming.Namespace;
import ru.mipt.acsl.decode.model.proxy.DecodeProxyResolver;

import java.util.List;

/**
 * @author Artem Shein
 */
class RegistryImpl implements Registry {

    private final ElementName name;
    private Namespace rootNamespace;
    private final List<DecodeProxyResolver> proxyResolvers;

    RegistryImpl(ElementName name, Namespace rootNamespace, List<DecodeProxyResolver> proxyResolvers) {
        this.name = name;
        this.rootNamespace = rootNamespace;
        this.proxyResolvers = proxyResolvers;

        if (Fqn.DECODE_NAMESPACE.size() != 1)
            throw new RuntimeException("not implemented");
    }

    @Override
    public Namespace rootNamespace() {
        return rootNamespace;
    }

    @Override
    public void setRootNamespace(Namespace ns) {
        rootNamespace = ns;
    }

    @Override
    public List<DecodeProxyResolver> proxyResolvers() {
        return proxyResolvers;
    }
}
